Lesson2

ilya-burinskiy opened the pull request.

Commit:
Models creation

Commit:
Added null constraint to model fields

- In Answer model question_id, user_id, answer_text, is_correct can not be null
- In Category model name can not be null
- In Question model question_txt, test can not be null
- In Test model title, level, category can not be null
- In User model first_name, last_name can not be null

Commit:
Default values

- Test level default value is 0
- Answer is_correct default value is false

Commit:
Deleted unnecessary migrations

Commit:
Model properties are defined in creation migration

VidgarVii commented on db/migrate/20201126090623_create_questions.rb:
```diff
@@ -0,0 +1,10 @@
+  def change
+    create_table :questions do |t|
+      t.text :question_txt, null: false
+      t.references :test, null: false
```
нужно указать foreign_key: true.

VidgarVii commented on db/migrate/20201126090623_create_questions.rb:
```diff
@@ -0,0 +1,10 @@
+class CreateQuestions < ActiveRecord::Migration[6.0]
+  def change
+    create_table :questions do |t|
+      t.text :question_txt, null: false
```
получилось тавтология. У тебя будет обращение question.question_txt. Что за txt? Почему 2 раза question?
Лучше обзови  пороше и понятней. title например

VidgarVii commented on db/migrate/20201126090028_create_tests.rb:
```diff
@@ -0,0 +1,11 @@
+    create_table :tests do |t|
+      t.string :title, null: false
+      t.integer :level, null: false
+      t.integer :category, null: false, default: 0
```
1 references
2 с default: 0 промахнулся

VidgarVii replied:
У теста должен быть автор

VidgarVii commented on db/migrate/20201126091258_create_answers.rb:
```diff
@@ -0,0 +1,12 @@
+      t.references :user, foreign_key: true, null: false
+      t.references :question, foreign_key: true, null: false
+      t.text :answer_text, null: false
+      t.boolean :is_correct, null: false, default: false
```
префикс is_ тут лишний. ОРМ для булевых полей предоставляет методы предикаты со знаком **?** `answer.correct? ` В руби приставка is_ не используется. А вот в JS нельзя в название функции использовать знак **?**. По этому они добавляют приставку is_

VidgarVii commented on db/migrate/20201126091258_create_answers.rb:
```diff
@@ -0,0 +1,12 @@
+class CreateAnswers < ActiveRecord::Migration[6.0]
+  def change
+    create_table :answers do |t|
+      t.references :user, foreign_key: true, null: false
```
Смотри на схему ЕРД. У ответов есть связь только с вопросами

VidgarVii commented on db/migrate/20201126091258_create_answers.rb:
```diff
@@ -0,0 +1,12 @@
+    create_table :answers do |t|
+      t.references :user, foreign_key: true, null: false
+      t.references :question, foreign_key: true, null: false
+      t.text :answer_text, null: false
```
про тавтологию указал выше.. При проектировании таблицы. Всегда думай как ты будешь ей пользоваться.

VidgarVii commented on db/migrate/20201126091629_create_users.rb:
```diff
@@ -0,0 +1,10 @@
+      t.string :last_name, null: false
+
+      t.timestamps
+    end
```
пока не критично. Если что потом по уроку добавишь. Но обычно пользователи имеют почту и пароль. А просто хранить имена людей, это как то.. странно

Commit:
Test model fixes

- level property has default value, not a category
- category is reference now

Commit:
Answer model fix

- Deleted reference to the User model
- answer_text renamed to the body
- is_correct renamed to correct?

Commit:
Question model fix

- question_txt renamed to title
- Added foreign_key: true to Test model reference

Commit:
User model fix

- Added email and password properties

VidgarVii commented on db/migrate/20201126091258_create_answers.rb:
```diff
@@ -0,0 +1,11 @@
+    create_table :answers do |t|
+      t.references :question, foreign_key: true, null: false
+      t.string :body, null: false
+      t.boolean :correct?, null: false, default: false
```
тут без вопросительного знака.

Commit:
Renamed correct? to correct in Answer model
